elide-core/src/main/java/com/yahoo/elide/core/EnforceJoinFilterExpressionVisitor.java

```diff
@@ -0,0 +1,106 @@
+ */
+public class EnforceJoinFilterExpressionVisitor implements FilterExpressionVisitor<Boolean> {
+    private PersistentResource<?> resource;
+
+    public EnforceJoinFilterExpressionVisitor(PersistentResource<?> resource) {
+        this.resource = resource;
+    }
+
+    /**
+     * Enforce ReadPermission on provided query filter
+     *
+     * @return true if allowed, false if rejected
+     */
+    @Override
+    public Boolean visitPredicate(FilterPredicate filterPredicate) {
+        RequestScope requestScope = resource.getRequestScope();
+        Set<PersistentResource> val = Collections.singleton(resource);
+        for (Path.PathElement pathElement : filterPredicate.getPath().getPathElements()) {
+            Class<?> entityClass = pathElement.getType();
+            String fieldName = pathElement.getFieldName();
+
+            if ("this".equals(fieldName)) {
+                continue;
+            }
+
+            try {
+                val = val.stream()
+                        .filter(Objects::nonNull)
+                        .flatMap(x -> getValueChecked(x, fieldName, requestScope))
+                        .filter(Objects::nonNull)
+                        .collect(Collectors.toSet());
+            } catch (IllegalArgumentException e) {
+                // Not a persistent resource
+            } catch (ForbiddenAccessException e) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private Stream<PersistentResource> getValueChecked(PersistentResource<?> resource, String fieldName,
+            RequestScope requestScope) {
+        // checkFieldAwareReadPermissions
+        requestScope.getPermissionExecutor().checkSpecificFieldPermissions(resource, null, ReadPermission.class,
+                fieldName);
+        Object obj = PersistentResource.getValue(resource.getObject(), fieldName, requestScope);
+        PersistentResourceSet persistentResourceSet;
+        if (obj instanceof Iterable) {
+            persistentResourceSet = new PersistentResourceSet(resource, (Iterable) obj, requestScope);
+        } else if (obj != null) {
+            persistentResourceSet = new PersistentResourceSet(resource, Collections.singleton(obj), requestScope);
+        } else {
+            return Stream.empty();
+        }
+
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(persistentResourceSet.iterator(), 0), false);
+    }
+
+    @Override
+    public Boolean visitAndExpression(AndFilterExpression expression) {
+        Boolean left = expression.getLeft().accept(this);
+        Boolean right = expression.getRight().accept(this);
+        // neither rejected
+        return left && right;
+    }
+
+    @Override
+    public Boolean visitOrExpression(OrFilterExpression expression) {
+        Boolean left = expression.getLeft().accept(this);
+        Boolean right = expression.getRight().accept(this);
+        // neither rejected
+        return left && right;
```

Comment:
should be left || right.

Reply:
No. true is valid, false is rejected.  If either is rejected, we want false.